src/hal/src/pso/mod.rs

```diff
@@ -221,32 +248,112 @@ pub struct SpecializationConstant {
 }
 
 /// Specialization information structure.
-#[derive(Debug, Copy)]
-pub struct Specialization<'a> {
+#[derive(Debug, Copy, Clone)]
+pub struct RawSpecialization<'a> {
     /// Constant array.
     pub constants: &'a [SpecializationConstant],
     /// Raw data.
     pub data: &'a [u8],
 }
 
-impl<'a> Default for Specialization<'a> {
+impl RawSpecialization<'_> {
+    /// Create empty specialization instance.
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+impl Default for RawSpecialization<'_> {
     fn default() -> Self {
-        Specialization {
+        RawSpecialization {
             constants: &[],
             data: &[],
         }
     }
 }
 
-impl<'a> Clone for Specialization<'a> {
-    fn clone(&self) -> Self {
-        Specialization {
-            constants: self.constants,
-            data: self.data,
+/// Convert typed specialization into untyped one.
+pub trait Specialization {
+    /// Build `Specialization` value.
+    fn as_raw(&self) -> RawSpecialization<'_>;
+}
+
+/// Temporary storage for specialization constants.
+#[derive(Clone, Debug, Default)]
+pub struct SpecializationStorage {
+    /// Actual data for specialization constants.
+    pub data: Vec<u8>,
+
+    /// Specialization constants that referer to the `data`.
+    pub constants: Vec<SpecializationConstant>,
+}
+
+impl SpecializationStorage {
+    /// Create empty specialization.
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+impl Specialization for SpecializationStorage {
+    fn as_raw(&self) -> RawSpecialization<'_> {
+        RawSpecialization {
+            data: &self.data,
+            constants: &self.constants,
         }
     }
 }
 
+/// Empty `Specialization` implementation.
+#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Unspecialized;
+
+impl Specialization for Unspecialized {
+    fn as_raw(&self) -> RawSpecialization<'_> {
+        RawSpecialization::new()
+    }
+}
+
+/// Trait for building specialization from list of constants.
+#[cfg(feature = "hlist-spec-const")]
+pub trait SpecializationList {
+    /// Fold list copying data into storage.
+    fn fold(&self, storage: &mut SpecializationStorage);
+
+    /// Build specialization from list.
+    fn build(&self) -> SpecializationStorage {
+        let mut storage = SpecializationStorage::new();
+        self.fold(&mut storage);
+        storage
+    }
+}
+
+#[cfg(feature = "hlist-spec-const")]
+impl SpecializationList for frunk_core::hlist::HNil {
+    fn fold(&self, _storage: &mut SpecializationStorage) {}
+}
+
+#[cfg(feature = "hlist-spec-const")]
+impl<H, T> SpecializationList for frunk_core::hlist::HCons<(u32, H), T>
```

Comment:
why don't we `use frunk_core::hlist` to begin with?

Reply:
Why do we need fancy `frunk_core::hlist` to begin with?